Transient CBS

NoyGabay opened the pull request:
add transient option for CBS,
main difference - use transient A*. 
CBS choosing the best node to expand - this happens naturally when giving CBS SumServiceTimes as a cost function

J-morag commented on src/main/java/BasicMAPF/Solvers/CBS/CBS_Solver.java:
```diff
@@ -178,7 +188,14 @@ private void initOpen(I_ConstraintSet initialConstraints) {
      */
     private CBS_Node generateRoot(I_ConstraintSet initialConstraints) {
-        Solution solution = new Solution(); // init an empty solution
+        Solution solution; // init an empty solution
+        if (transientMAPFBehaviour.isTransientMAPF()) {
+            solution = new TransientMAPFSolution();
+        }
+        else {
+            solution = new Solution();
+        }
+
```
```suggestion
        // init an empty solution
        Solution solution = transientMAPFBehaviour.isTransientMAPF() ? new TransientMAPFSolution() : new Solution(); 
```
just a little cleaner

J-morag commented on src/main/java/BasicMAPF/Solvers/CBS/CBS_Solver.java:
```diff
@@ -281,7 +298,15 @@ private CBS_Node generateNode(CBS_Node parent, Constraint constraint, boolean co
         // replace with copies if required
         if(copyDatastructures) {
-            solution = new Solution(solution);
+//            solution = new Solution(solution);
+
+            if (transientMAPFBehaviour.isTransientMAPF()) {
+                solution = new TransientMAPFSolution(solution);
+            }
+            else {
+                solution = new Solution(solution);
+            }
+
```
```suggestion
            solution =  transientMAPFBehaviour.isTransientMAPF() ? new TransientMAPFSolution(solution) : new Solution(solution);
```
a little cleaner

J-morag commented on src/main/java/Environment/RunManagers/GenericRunManager.java:
```diff
@@ -49,10 +50,16 @@ void setSolvers() {
-                new RestartsStrategy(RestartsStrategy.RestartsKind.none, 0, RestartsStrategy.RestartsKind.randomRestarts),
-                null, null, null));
-        super.solvers.add(new CBS_Solver());
+//        super.solvers.add(new PrioritisedPlanning_Solver(null, null, null,
+//                new RestartsStrategy(RestartsStrategy.RestartsKind.none, 0, RestartsStrategy.RestartsKind.randomRestarts),
+//                null, null, null));
+//        super.solvers.add(new PrioritisedPlanning_Solver(null, null, null,
+//                new RestartsStrategy(RestartsStrategy.RestartsKind.none, 0, RestartsStrategy.RestartsKind.randomRestarts),
+//                null, null, TransientMAPFBehaviour.transientMAPF));
+        super.solvers.add(new CBS_Solver(null, null, null, null, null, null, null, null, null));
+        super.solvers.add(new CBS_Solver(null, null, null, null, null, null, null, null, TransientMAPFBehaviour.transientMAPF));
+        super.solvers.add(new CBS_Solver(null, null, null, new SumServiceTimes(), null, null, null, null, TransientMAPFBehaviour.transientMAPFsstWithBlacklist));
+
```
```suggestion
        super.solvers.add(new PrioritisedPlanning_Solver(null, null, null,
                new RestartsStrategy(RestartsStrategy.RestartsKind.none, 0, RestartsStrategy.RestartsKind.randomRestarts),
                null, null, null));
        super.solvers.add(new CBS_Solver());
```

J-morag commented on src/test/java/BasicMAPF/Solvers/CBS/CBS_SolverTest.java:
```diff
@@ -416,4 +418,27 @@ void sharedGoals(){
         }
     }
+
+    @Test
+    void worksWithTMAPFPaths() {
+        I_Solver CBSt = new CBS_Solver(null, null, null, null, null, null, null, null, TransientMAPFBehaviour.transientMAPF);
+        Agent agentXMoving = new Agent(0, coor42, coor02, 1);
+        Agent agentYMoving = new Agent(1, coor10, coor12, 1);
+        MAPF_Instance testInstance = new MAPF_Instance("testInstance", mapEmpty, new Agent[]{agentXMoving, agentYMoving});
+
+        Solution solvedNormal = cbsSolver.solve(testInstance, new RunParametersBuilder().setTimeout(1000L).setInstanceReport(instanceReport).createRP());
+        assertTrue(solvedNormal.solves(testInstance));
+        assertEquals(4 + 4, solvedNormal.sumIndividualCosts());
+        assertEquals(6, solvedNormal.makespan());
+
+        Solution solvedCBSt = CBSt.solve(testInstance, new RunParametersBuilder().setTimeout(1000L).setInstanceReport(instanceReport).createRP());
+        assertTrue(solvedCBSt.solves(testInstance));
+        assertEquals(4 + 3, solvedCBSt.sumIndividualCosts()); // normal SOC function
+        assertEquals(4 + 2, solvedCBSt.sumServiceTimes()); // TMAPF cost function
+        assertEquals(4, solvedCBSt.makespan()); // makespan (normal)
+        assertEquals(4, solvedCBSt.makespanServiceTime()); // makespan (TMAPF)
+
+        System.out.println(solvedNormal);
+        System.out.println(solvedCBSt);
+    }
```
Can you also add the example from the paper?

J-morag replied:
Use a `MapFactory.newArbitraryGraphMap` for it because it isn't on a grid

NoyGabay replied:
Done :)

J-morag reviewed (approved):
A few small comments. good job :)

NoyGabay commented:
Added the example from the paper - looks good. 
I'm wondering if I implemented transientMAPFWithBlacklist correctly in CBS, since I didn't distinguished between it and transientMAPF in the code. transientMAPFsstWithBlacklist acts like them, excluding the sst optimization heuristic.

J-morag reviewed (approved):
For 'transientMAPFWithBlacklist', It should blacklist the targets of all other agents, whenever planning one agent. You can add that, test it, and merge. It would make CBS incomplete, and might not be the best thing to do in CBS, but we need to run experiments and see.
For 'transientMAPFsstWithBlacklist' - basically we have a duplication of indicating sst here and in the cost function parameter. We need to do some refactoring. At the very least we need to make the 'enum' a 'record' to separate out the parameters and make any combination possible. We should probably also forgo having SST indicated there, and instead only use the costFunction interface. It's a bit of a mess. But finish up this PR, and then we can refactor it to be cleaner in another PR.
